implemented the ThingTypeMigrator service

afuechsel opened the pull request:
Signed-off-by: Andre Fuechsel <andre.fuechsel@telekom.de>

tomhoefer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingTypeMigrator.java:
```diff
@@ -0,0 +1,29 @@
+     * @param thingTypeUID new {@link ThingTypeUID}
+     * @param configuration new configuration
+     */
+    void changeThingType(Thing thing, ThingTypeUID thingTypeUID, Configuration configuration);
```
Would it make sense to rename the operation to migrateThingType if the service is called migrator?

afuechsel replied:
Hmm, it doesn't really migrate, it just changes, but ThingTypeChanger as service name sounds odd to me. So yes, I could rename it to migrateThingType

tomhoefer replied:
sure... what about ThingTypeConverter with operation convert?

afuechsel replied:
After a second thought I would stick to ThingTypeMigrator#migrateThingType

sjka replied:
I also like "migration" better, as it highlights the aspect of moving over instead of just changing a value somewhere.

tomhoefer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -374,6 +306,91 @@ public void handlerRemoved(Thing thing, ThingHandler thingHandler) {
+                        Thread.sleep(100);
+                        logger.debug(
+                                "Waiting for handler deregistration to complete for thing {}. Took already {}ms.",
+                                thing.getUID().getAsString(), i * 100);
```
This log is wrong (in general we must either count from i =1 or increment i +1 here to get the correct time)

afuechsel replied:
You are right, this was also wrong in the original code.

tomhoefer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -247,78 +250,7 @@ public void configurationUpdated(Thing thing) {
         @Override
         public void changeThingType(final Thing thing, final ThingTypeUID thingTypeUID,
```
if we decide to rename changeThingType to migrateThingType for ThingTypeMigrator then we should think about to rename this operation as well

tomhoefer replied:
if we decide to rename changeThingType to migrateThingType for ThingTypeMigrator then we should think about to rename this operation as well

tomhoefer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -247,78 +250,7 @@ public void configurationUpdated(Thing thing) {
-                            thingUID.getAsString()));
-                }
-            }, 0, TimeUnit.MILLISECONDS);
+            ThingManager.this.changeThingType(thing, thingTypeUID, configuration);
```
if we decide to rename changeThingType to migrateThingType for ThingTypeMigrator then we should think about to rename the callback operation as well

tomhoefer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -374,6 +306,91 @@ public void handlerRemoved(Thing thing, ThingHandler thingHandler) {
+                ThingUID thingUID = thing.getUID();
+                waitForRunningHandlerRegistrations(thingUID);
+                ThingType thingType = thingTypeRegistry.getThingType(thingTypeUID);
+                if (thingType == null) {
```
i would prefer if we check this outside the runnable implementation and throw an exception if the thing type does not exist... then we should document this exception for the service and for the callback

afuechsel replied:
yes, can be changed

tomhoefer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -374,6 +306,91 @@ public void handlerRemoved(Thing thing, ThingHandler thingHandler) {
+                thingRegistry.update(thing);
+    
+                logger.debug("Changed ThingType of Thing {} to {}. New ThingHandler is {}.",
+                        thing.getUID().toString(), thing.getThingTypeUID(), thing.getHandler().toString());
```
is it ensured that the handler is already set here (to avoid NPE)?

afuechsel replied:
No, it is not. Needs to be changed.

afuechsel replied:
Second thought: yes, it is, the new handler is registered just the line before. No need to change.

sjka replied:
right, due to a nearly endless chain of luckily synchronous calls, this currently is guaranteed.

sjka commented:
Overall, from my point of view introducing the interface make sense in order to allow bindings (or others...) to change the ThingType independently of the ThingHandler.

sjka commented:
lgtm

beware: the signature change ~~changeThingType~~ -> migrateThingType in ThingHandlerCallback is incompatible. As there are very few usages so far, imho this is acceptable, but at least @cdjackson should be made aware of it.

afuechsel commented:
@sjka Shall we revoke the change in ThingHandlerCallback? Back to changeThingType?

afuechsel commented:
I have no clue, why there are so many test failures. Executing the tests locally, all were green

tomhoefer commented:
I would not revert the change in the callback interface because I guess there are only a few bindings using this feature

kaikreuzer commented:
Ok for me as well to change the name - I think we can cope with adapting the bindings that use it (@cdjackson ok for you?).

I am rather wondering if we can find a better name for the "Migrator" - this does not sound to me like a service, but rather like some job/task to instantiate and run. Maybe simply "ThingTypeMigrationService" instead?

Do we need any changes to the documentation for this?

cdjackson commented:
Ok for me to change it, but this will be a breaking change in ZWave binding, so we should coordinate the ESH stable -> OH2 build with the required update in ZWave...

afuechsel commented:
@kaikreuzer I originally thought about ThingTypeMigrationService, but wanted to avoid the term Service, as all other services also are not called ...Service but I agree it sounds better in this particular case. Will change it. 

@cdjackson you should also be aware of one slight small bug fix: originally the implementation did not check, whether or not the new thing type is registered in the thing type registry. This led to the effect, that in the case, the thing type was *not* in the registry, that the respective thing was left in UNITIALIZED status but the thing type unchanged. Just after a handler restrt (calling initialize() of the thing again) the thing was ONLINE again. I added a check just before the asynchronous change process throwing a RuntimeException in the case the thing type is not in the thing type registry (documented on the interface). 

I will make the final changes tomorrow and hope the build is green because currently I see no reason for the failures in the build.

cdjackson commented:
That won’t be an issue for ZWave - it should only ever change the thing type to something that it knows is in the registry since it checks the registry first to compare some properties to find the thing to change to...

afuechsel commented:
Implemented all changes and fixed the tests (the reason of the failing tests was, that changeThingType tests actually changed the thing type of the static test object THING leading other tests to fail...)

kaikreuzer commented:
Thanks.
@cdjackson I will let you know when this flows to OH - possibly tomorrow if this is fine for you.

cdjackson commented:
That's fine with me - just ping me when you confirm...

kaikreuzer commented:
@cdjackson Hoping that the build is green again today, I would try to do a new stable release tonight - I will then simply deactivate the Z-Wave build plan, so that you can do the necessary adaptions and simply enable it again afterwards.

cdjackson commented:
Ok - that’s fine with me - thanks.

cdjackson commented:
I didn't read this properly. Since this is a change to the ThingHandlerCallback, and not the ThingHandler, it doesn't matter to the binding - it just needs recompiling.

kaikreuzer commented:
Ok, then you do not have to worry. But the stable for tonight is cancelled, since the build is still not back to normal (see https://github.com/eclipse/smarthome/issues/2003) 😩
